### PR TITLE
Fix sync error in the face of renames and duplicate UUIDs

### PR DESCRIPTION
**Pull request by nwalters512**
Fixes https://github.com/PrairieLearn/PrairieLearn/issues/6539. See inline comments for details on the fix itself! Since we use essentially the same syncing algorithm for questions, assessments, and course instances, I had to make the fix in all three places.

### CODE REVIEW

File: tests/sync/questionsSync.test.js [hunk 1]
```diff
@@ -358,4 +358,26 @@ describe('Question syncing', () => {
     assert.equal(deletedQuestion.uuid, originalQuestion.uuid);
     assert.equal(deletedQuestion.sync_errors, null);
   });
+
+  it.only('reproduces problem Firas reported', async () => {
```
Comment (nwalters512) [inline tests/sync/questionsSync.test.js:362]: 👀 drop `.only` before merging.

File: tests/sync/util.js [hunk 2]
```diff
@@ -269,6 +269,7 @@ module.exports.writeCourseToDirectory = async function (courseData, coursePath)
 module.exports.QUESTION_ID = 'test';
 module.exports.ALTERNATIVE_QUESTION_ID = 'test2';
 module.exports.MANUAL_GRADING_QUESTION_ID = 'test_manual';
+module.exports.WORKSPACE_QUESTION_ID = 'workspace';
```
Comment (nwalters512) [inline tests/sync/util.js:272]: This was missing; I only noticed after I saw a question with the qid `undefined` while debugging some tests.

File: tests/sync/util.js [hunk 3]
```diff
@@ -72,6 +72,16 @@ const syncFromDisk = require('../../sync/syncFromDisk');
  * @property {"Set" | "Module"} [groupAssessmentsBy]
  */
 
+/**
```
Comment (nwalters512) [inline tests/sync/util.js:75]: I turned on type checking for `tests/sync/assessmentsSync.test.js`, which necessitated updating a lot of things here.

Comment (mwest1066) [reply tests/sync/util.js:75]: Nice!

**Review by mwest1066 — APPROVED** [review]
Wow, such a sneaky bug! Thanks for the clear explanations in the linked issue and in the new comments in the code. This makes total sense with the explanation.

**Pull request merged**